sunset/code/modules/vending/_vending.dm

```diff
@@ -0,0 +1,569 @@
+		return FALSE
+	if(!component_parts || !refill_canister)
+		return FALSE
+
+	var/moved = 0
+	if(panel_open || W.works_from_distance)
+		if(W.works_from_distance)
+			display_parts(user)
+		for(var/I in W)
+			if(istype(I, refill_canister))
+				moved += restock(I)
+	else
+		display_parts(user)
+	if(moved)
+		to_chat(user, "[moved] items restocked.")
+		W.play_rped_sound()
+	return TRUE
+
+/obj/machinery/vending/on_deconstruction()
+	update_canister()
+	. = ..()
+
+/obj/machinery/vending/emag_act(mob/user)
+	if(obj_flags & EMAGGED)
+		return
+	obj_flags |= EMAGGED
+	to_chat(user, "<span class='notice'>You short out the product lock on [src].</span>")
+
+/obj/machinery/vending/_try_interact(mob/user)
+	if(seconds_electrified && !(stat & NOPOWER))
+		if(shock(user, 100))
+			return
+	return ..()
+
+/obj/machinery/vending/ui_interact(mob/user)
+	var/dat = ""
+
+	dat += "<h3>Select an item</h3>"
+	dat += "<div class='statusDisplay'>"
+	if(!product_records.len)
+		dat += "<font color = 'red'>No product loaded!</font>"
+	else
+		var/list/display_records = product_records
+		if(extended_inventory)
+			display_records = product_records + hidden_records
+		if(coin || bill)
+			display_records = product_records + coin_records
+		if((coin || bill) && extended_inventory)
+			display_records = product_records + hidden_records + coin_records
+		dat += "<table>"
+		for (var/datum/data/vending_product/R in display_records)
+			dat += "<tr><td>" + GetIconForProduct(R) + "</td>"
+			dat += "<td style=\"width: 100%\"><b>[sanitize(R.name)]</b></td>"
+			if(R.amount > 0)
+				dat += "<td><b>[R.amount]&nbsp;</b></td><td><a href='byond://?src=\ref[src];vend=\ref[R]'>Vend</a></td>"
+			else
+				dat += "<td>0&nbsp;</td><td><span class='linkOff'>Vend</span></td>"
+				dat += "</tr>"
+				dat += "</table>"
+	dat += "</div>"
+	if(premium.len > 0)
+		dat += "<b>Change Return:</b> "
+		if (coin || bill)
+			dat += "[(coin ? coin : "")][(bill ? bill : "")]&nbsp;&nbsp;<a href='byond://?src=[REF(src)];remove_coin=1'>Remove</a>"
+		else
+			dat += "<i>No money</i>&nbsp;&nbsp;<span class='linkOff'>Remove</span>"
+	if(istype(src, /obj/machinery/vending/snack))
+		dat += "<h3>Chef's Food Selection</h3>"
+		dat += "<div class='statusDisplay'>"
+		for (var/O in dish_quants)
+			if(dish_quants[O] > 0)
+				var/N = dish_quants[O]
+				dat += "<a href='byond://?src=[REF(src)];dispense=[sanitize(O)]'>Dispense</A> "
+				dat += "<B>[capitalize(O)]: [N]</B><br>"
+		dat += "</div>"
+
+	var/datum/browser/popup = new(user, "vending", (name))
+	popup.set_content(dat)
+	popup.set_title_image(user.browse_rsc_icon(icon, icon_state))
+	popup.open()
+
+GLOBAL_LIST_EMPTY(vending_machine_icon)
+/obj/machinery/vending/proc/GetIconForProduct(datum/data/vending_product/P)
+	var/producticon = GLOB.vending_machine_icon[P.product_path]
+	if (producticon)
+		return producticon
+	var/product = new P.product_path()
+	producticon = icon2base64html(product)
```

Comment:
you should use `icon2base64(getFlatIcon(product, no_anim = TRUE))` to cover items with overlays, as well as animated icons

Reply:
I'm still working on it, but this is what happens if I just swap that out:
![change of geticon](https://user-images.githubusercontent.com/43192081/46056224-9ef70000-c11d-11e8-8768-f5280e9d132c.png)

Reply:
also this: `<img src='data:image/jpeg;base64,[GetIconForProduct(R)]'/>`